More typings

ostcar opened the pull request.

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -92,6 +94,12 @@ def ws_add_site(message):
         access_permissions = collection.get_access_permissions()
         restricted_data = access_permissions.get_restricted_data(collection, user)
 
+        if restricted_data is None:
```
This can should never happen: In line 93 (old) / 95 (new) we get the restricted data using a Collection And not a single instance. In this case all restricted_data functions have to return an iterable (e. g. an empty list).

Is it possible to remove the two lines?

ostcar replied:
Yes, it is possible. The cast() call in the next line should solve this. But please be aware of this issue: https://github.com/OpenSlides/OpenSlides/issues/3377

My goal is still to remove the "container" object in the get_restricted_data methods and replace it with List[Dict[str, Any]] (maybe the value in the dict could also be str, or Unique[str, int, bool])

normanjaeckel replied:
Issue #3377 is fixed in current master, so please remove these lines.

ostcar replied:
done

normanjaeckel commented on openslides/utils/access_permissions.py:
```diff
@@ -1,46 +1,35 @@
 
+from .collection import Collection, CollectionElement
 
-class BaseAccessPermissions(object, metaclass=SignalConnectMetaClass):
```
Why do you remove the metaclass here?  Maybe because the feature is not used any more ...?

ostcar replied:
Yes. And I look forward of removing utils.dispatch anywhere. I think our new style, to explicit add objects to a list in app.raddy is match better.

normanjaeckel commented on openslides/utils/access_permissions.py:
```diff
@@ -1,46 +1,35 @@
+class BaseAccessPermissions:
     """
-    Base access permissions container.
+    Base access permissions containser.
```
-1

ostcar replied:
done

normanjaeckel commented on openslides/utils/access_permissions.py:
```diff
@@ -1,46 +1,35 @@
-    signal = Signal()
 
-    def __init__(self, **kwargs):
+    def __init__(self) -> None:
```
The init method can be removed here.

ostcar replied:
done

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -64,7 +66,7 @@ def format_for_autoupdate(collection_string, id, action, data=None):
 
 @channel_session_user_from_http
-def ws_add_site(message):
+def ws_add_site(message: Any) -> None:
```
Message is a specific type, isn't it?

ostcar replied:
It is a Dict. Maybe you can use Dict[str, Any]. I don't belief that we can set an explicit type for the Value.

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -16,7 +18,7 @@
 
 
-def send_or_wait(send_func, *args, **kwargs):
+def send_or_wait(send_func: Any, *args: Any, **kwargs: Any) -> None:
```
send_func should always be a channel's send function. Args and kwargs should only be the args and kwargs for such a function.

ostcar replied:
I did not understand all of the functions you write. Please add a commit to the pull request where you add as match typings as possible.

It should be the goal, that there is no Any in the code. (But we will probably not  succeed.)

normanjaeckel commented on openslides/utils/cache.py:
```diff
@@ -252,7 +269,7 @@ def use_redis_cache():
 
 
-def get_redis_connection():
+def get_redis_connection() -> Any:
```
Hm. Is it possible to set the correct type easily?

ostcar replied:
I don't know what type it is.

normanjaeckel commented on setup.cfg:
```diff
@@ -18,7 +18,10 @@ multi_line_output = 3
 
-[mypy-openslides.utils.auth]
+[mypy-openslides.utils.dispatch]
+ignore_errors = true
```
Which errors remain?

ostcar replied:
This file is a bit complicated and I do not understand it. As I wrote above, I would like to remove it. Therefore I don't want to spend any time on it.
But of cause, you can add typing information as you like

ostcar commented:
All errors done. Please merge this. It has no new feature. If you like to add more typings you can do this in another pull request.

ostcar commented:
@normanjaeckel I added the option ```check_untyped_defs``` to check code outside of utils. This option can be removed when the entire project has type annotations. I added one ```#TODO``` (without the space) with a question. What should we do in this case?

normanjaeckel commented on openslides/utils/autoupdate.py:
```diff
@@ -92,6 +94,12 @@ def ws_add_site(message):
+            continue
+
+        # At this point restricted_data has to be a list. So we have to tell it mypy
+        restricted_data = cast(List[Dict[str, Any]], restricted_data)
```
Please check if this is still required.

ostcar replied:
Yes it is. mypy can not know, that at this point restricted_data is a list. ```cast()``` is there for situations like this.

ostcar commented:
All done and rebased. I merge after travis is ok
